nntrainer/nntrainer_error.h

```diff
@@ -24,13 +24,72 @@
 #define ML_ERROR_RESULT_OUT_OF_RANGE (-ERANGE)
 #endif
 
+#include <functional>
+#include <sstream>
 #include <stdexcept>
+
+#define NNTR_THROW_IF(pred, err) \
+  if (!(pred)) {                 \
+  } else                         \
+    nntrainer::exception::internal::ErrorNotification<err>()
```

Comment:
This looks extremely dangerious. Can you make the macro closed?

E.g.,
```
#define NNTR_THROW_IF(pred, err, message) \
   do { \
      ....
   } while (0)
```

Reply:
Note that the following code works. So don't worry:
```
#include <iostream>
#define A(a)\
        do { \
                std::cout << a << std::endl; \
        } while (0)

int main()
{
        A(123123 << std::endl << "33" << std::endl);
}
```

Reply:
Question:
  Why do you need the empty "then" statement here? Are you trying to hack the branch-prediction mechanism?

Reply:
Not closing the stream intentional. This enables logging directly to a stream, as [gtest](https://stackoverflow.com/a/16491892) or [aosp](https://cs.android.com/android/platform/superproject/+/master:packages/modules/NeuralNetworks/runtime/CompilationBuilder.cpp;l=103) does.

They also do not close the stream at the end.

I tried to minimize the code that needs to be inside the macro. Similar to implementation in [aosp](https://cs.android.com/android/platform/superproject/+/master:system/libbase/include/android-base/logging.h;drc=master;l=219) is doing.
If it doesn't sound right, I will change to the suggested style.

Reply:
> Question:
> Why do you need the empty "then" statement here? Are you trying to hack the branch-prediction mechanism?

It is not necessary, I'll change.

Reply:
> > Question:
> > Why do you need the empty "then" statement here? Are you trying to hack the branch-prediction mechanism?
> 
> It is not necessary, I'll change.

No. keep as it is. If you are trying their style, Having an empty "then statement" is safer. Assume you have :
```
if (some condition) NNTR_THROW_IF (X, Y) << "message";
else NNTR_THROW_IF (Z, W) << "message2";
```

Reply:
Keep the original code if your intention is to follow their style. The whole statement is a single C statement; thus, do-while is not really required.

Reply:
Awesome :)